### PR TITLE
Removed mutex thread locking and unlocking from logging Format registry

### PR DESCRIPTION
**Pull request by rahulchheda**
Signed-off-by: Rahul M Chheda <rchheda@infracloud.io>

**What type of PR is this?**

> /kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #https://github.com/kubernetes/kubernetes/issues/92346

**Special notes for your reviewer**:

ref: https://github.com/kubernetes/kubernetes/pull/92177#discussion_r442897848

**Does this PR introduce a user-facing change?**:

```release-note

NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by serathius** [comment]
/lgtm
Please follow instructions https://github.com/kubernetes/community/blob/master/CLA.md#the-contributor-license-agreement to sign CLA

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 1]
```diff
@@ -33,21 +32,17 @@ var logRegistry = NewLogFormatRegistry()
...
 		registry: make(map[string]logr.Logger),
-		mu:       sync.Mutex{},
 	}
 }
 
 // Register new log format registry to global logRegistry
 func (lfr *LogFormatRegistry) Register(name string, logger logr.Logger) error {
-	lfr.mu.Lock()
-	defer lfr.mu.Unlock()
 	if _, ok := lfr.registry[name]; ok {
```
Comment (cheftako) [inline staging/src/k8s.io/component-base/logs/registry.go:46]: I'm all for the locks being removed. However I believe part of the reasoning for removing the locks was that you can't reasonably add new loggers after initialization (or more specifically after validation). I think the API should reflect that. My preference would be to make Set/Register private. I know that lavalamp mentioned making them blow up if called after validation. (I would be fine with that as well)

Comment (lavalamp) [reply staging/src/k8s.io/component-base/logs/registry.go:46]: Yeah it seems unlikely that people will accidentally make initialization multi-threaded.

Accidentally registering post-validate / flag generation seems like a more likely error, I think it'd be reasonable to set a boolean and panic in that condition.

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 2]
```diff
@@ -32,21 +32,30 @@ var logRegistry = NewLogFormatRegistry()
...
 func NewLogFormatRegistry() *LogFormatRegistry {
 	return &LogFormatRegistry{
 		registry: make(map[string]logr.Logger),
+		freezed:  false,
 	}
 }
 
 // Register new log format registry to global logRegistry
 func (lfr *LogFormatRegistry) Register(name string, logger logr.Logger) error {
+	if lfr.freezed {
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/registry.go:48]: I think we should protect all mutations, please add this to Set and Delete

Comment (rahulchheda) [reply staging/src/k8s.io/component-base/logs/registry.go:48]: @serathius , got it.
Btw, any changes needed in error string?

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 3]
```diff
@@ -32,21 +32,30 @@ var logRegistry = NewLogFormatRegistry()
...
+		return fmt.Errorf("log format is freezed, unable to register log format")
+	}
 	if _, ok := lfr.registry[name]; ok {
 		return fmt.Errorf("log format: %s already exists", name)
 	}
 	lfr.registry[name] = logger
+
+	// Freeze log format registry
+	// to make it unable to registy new log formats
+	lfr.freezed = true
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/registry.go:58]: I think feezing should be done after --logging-format flag calls Apply, you can add Freeze function to Register and call it in Apply https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/logs/options.go#L69

File: staging/src/k8s.io/component-base/logs/options.go [hunk 4]
```diff
@@ -63,13 +63,16 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 	unsupportedFlags := fmt.Sprintf("--%s", strings.Join(unsupportedLoggingFlags(), ", --"))
 	formats := fmt.Sprintf(`"%s"`, strings.Join(logRegistry.List(), `", "`))
 	fs.StringVar(&o.LogFormat, logFormatFlagName, defaultLogFormat, fmt.Sprintf("Sets the log format. Permitted formats: %s.\nNon-default formats don't honor these flags: %s.\nNon-default choices are currently alpha and subject to change without warning.", formats, unsupportedFlags))
+
+	// Frezze Log Registry after setting logger
+	// which make it unable to register new logger
+	logRegistry.freezed = true
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/options.go:69]: All other functions used from registry are public, can you add function `Freeze` and call it here?

File: staging/src/k8s.io/component-base/logs/options.go [hunk 5]
```diff
@@ -63,13 +63,16 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 	unsupportedFlags := fmt.Sprintf("--%s", strings.Join(unsupportedLoggingFlags(), ", --"))
 	formats := fmt.Sprintf(`"%s"`, strings.Join(logRegistry.List(), `", "`))
 	fs.StringVar(&o.LogFormat, logFormatFlagName, defaultLogFormat, fmt.Sprintf("Sets the log format. Permitted formats: %s.\nNon-default formats don't honor these flags: %s.\nNon-default choices are currently alpha and subject to change without warning.", formats, unsupportedFlags))
+
+	// Frezze Log Registry after setting logger
+	// which make it unable to register new logger
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/options.go:68]:
```suggestion
	// No new log formats should be added after generation is of flag options
```

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 6]
```diff
@@ -33,21 +32,22 @@ var logRegistry = NewLogFormatRegistry()
 // LogFormatRegistry store klog format registry
 type LogFormatRegistry struct {
 	registry map[string]logr.Logger
-	mu       sync.Mutex
+	freezed  bool
```
Comment (cheftako) [inline staging/src/k8s.io/component-base/logs/registry.go:35]: frozen

Comment (rahulchheda) [reply staging/src/k8s.io/component-base/logs/registry.go]: fixed. PTAL

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 7]
```diff
@@ -33,21 +32,22 @@ var logRegistry = NewLogFormatRegistry()
...
+		frozen:   false,
 	}
 }
 
 // Register new log format registry to global logRegistry
 func (lfr *LogFormatRegistry) Register(name string, logger logr.Logger) error {
-	lfr.mu.Lock()
-	defer lfr.mu.Unlock()
+	if lfr.frozen {
+		return fmt.Errorf("log format is freezed, unable to register log format")
```
Comment (cheftako) [inline staging/src/k8s.io/component-base/logs/registry.go:49]: is frozen

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 8]
```diff
@@ -67,30 +65,38 @@ func (lfr *LogFormatRegistry) Get(name string) (logr.Logger, error) {
 }
 
 // Set specified log format logger
-func (lfr *LogFormatRegistry) Set(name string, logger logr.Logger) {
-	lfr.mu.Lock()
-	defer lfr.mu.Unlock()
+func (lfr *LogFormatRegistry) Set(name string, logger logr.Logger) error {
+	if lfr.frozen {
+		return fmt.Errorf("log format is freezed, unable to set log format")
```
Comment (cheftako) [inline staging/src/k8s.io/component-base/logs/registry.go:70]: ditto

File: staging/src/k8s.io/component-base/logs/registry.go [hunk 9]
```diff
@@ -67,30 +65,38 @@ func (lfr *LogFormatRegistry) Get(name string) (logr.Logger, error) {
...
+	return nil
 }
 
 // Delete specified log format logger
-func (lfr *LogFormatRegistry) Delete(name string) {
-	lfr.mu.Lock()
-	defer lfr.mu.Unlock()
+func (lfr *LogFormatRegistry) Delete(name string) error {
+	if lfr.frozen {
+		return fmt.Errorf("log format is freezed, unable to delete log format")
```
Comment (cheftako) [inline staging/src/k8s.io/component-base/logs/registry.go:80]: ditto

Comment (rahulchheda) [reply staging/src/k8s.io/component-base/logs/registry.go]: done, sorry for delaying the PR with these small changes.

**Comment by serathius** [comment]
Can you squash commits?

**Pull request merged**